File: go/api/item.go
package api

import (
	"encoding/json"
	"fmt"
)

type Inventory struct {
	ExtraColumns int `json:"extra_columns"`
	// 金币
	Gold int `json:"gold"`
}

// Influences 影响
type Influences struct {
	Redeemer *bool `json:"redeemer,omitempty"`
	Shaper   *bool `json:"shaper,omitempty"`
	Elder    *bool `json:"elder,omitempty"`
	Crusader *bool `json:"crusader,omitempty"`
	Hunter   *bool `json:"hunter,omitempty"`
	Warlord  *bool `json:"warlord,omitempty"`
}

// SocketAttribute 插槽属性
//
// I: Int/智慧, D: Dex/敏捷, S: Str/力量, G: General/通用, A: Abyssal/深渊
type SocketAttribute string

const (
	SocketAttributeInt     SocketAttribute = "I"
	SocketAttributeDex     SocketAttribute = "D"
	SocketAttributeStr     SocketAttribute = "S"
	SocketAttributeGeneral SocketAttribute = "G"
	SocketAttributeAbyssal SocketAttribute = "A"
)

// SocketColour 插槽颜色
//
// B: Blue/蓝色, G: Green/绿色, R: Red/红色, W: White/白色, A: Abyssal/深渊
type SocketColour string

const (
	SocketColourBlue    SocketColour = "B"
	SocketColourGreen   SocketColour = "G"
	SocketColourRed     SocketColour = "R"
	SocketColourWhite   SocketColour = "W"
	SocketColourAbyssal SocketColour = "A"
)

// Rarity 稀有度
type Rarity string

const (
	RarityNormal Rarity = "Normal"
	RarityMagic  Rarity = "Magic"
	RarityRare   Rarity = "Rare"
	RarityUnique Rarity = "Unique"
)

// Socket 插槽
type Socket struct {
	Group   int             `json:"group"`
	Attr    SocketAttribute `json:"attr"`
	SColour SocketColour    `json:"sColour"`
}

// Value Requirements/Properties等类型的值
//
// 其 JSON 表示为 [string, int] 混合数组
type Value struct {
	Value string `json:"-"`
	Index int    `json:"-"`
}

func (v Value) MarshalJSON() ([]byte, error) {
	return json.Marshal([]any{v.Value, v.Index})
}

func (v *Value) UnmarshalJSON(data []byte) error {
	var arr [2]any
	if err := json.Unmarshal(data, &arr); err != nil {
		return err
	}

	if str, ok := arr[0].(string); ok {
		v.Value = str
	} else {
		return fmt.Errorf("expected string at index 0, got %T", arr[0])
	}

	if num, ok := arr[1].(float64); ok {
		v.Index = int(num)
	} else if num, ok := arr[1].(int); ok {
		v.Index = num
	} else {
		return fmt.Errorf("expected number at index 1, got %T", arr[1])
	}

	return nil
}

// Requirement 需求
type Requirement struct {
	Name        string  `json:"name"`
	Values      []Value `json:"values"`
	DisplayMode int     `json:"displayMode"`
	Type        *int    `json:"type,omitempty"`
	Suffix      *string `json:"suffix,omitempty"`
}

// Property 属性
type Property struct {
	Name        string   `json:"name"`
	Values      []Value  `json:"values"`
	DisplayMode int      `json:"displayMode"`
	Type        *int     `json:"type,omitempty"`
	Progress    *float64 `json:"progress,omitempty"`
}

// Item 物品
type Item struct {
	// 仅限：深渊珠宝
	// 深渊珠宝必有该属性且为true
	AbyssJewel    *bool    `json:"abyssJewel,omitempty"`
	BaseType      string   `json:"baseType"`
	Corrupted     *bool    `json:"corrupted,omitempty"`
	CraftedMods   []string `json:"craftedMods,omitempty"`
	CrucibleMods  []string `json:"crucibleMods,omitempty"`
	DescrText     *string  `json:"descrText,omitempty"`
	Duplicated    *bool    `json:"duplicated,omitempty"`
	Elder         *bool    `json:"elder,omitempty"`
	EnchantMods   []string `json:"enchantMods,omitempty"`
	ExplicitMods  []string `json:"explicitMods,omitempty"`
	FlavourText   []string `json:"flavourText,omitempty"`
	FoilVariation *int     `json:"foilVariation,omitempty"`
	Fractured     *bool    `json:"fractured,omitempty"`
	FracturedMods []string `json:"fracturedMods,omitempty"`
	FrameType     int      `json:"frameType"`
	H             int      `json:"h"`
	// 仅限：宝石
	Hybrid        *Hybrid        `json:"hybrid,omitempty"`
	Icon          string         `json:"icon"`
	ID            *string        `json:"id,omitempty"`
	Identified    bool           `json:"identified"`
	Ilvl          int            `json:"ilvl"`
	ImplicitMods  []string       `json:"implicitMods,omitempty"`
	Influences    *Influences    `json:"influences,omitempty"`
	InventoryId   *string        `json:"inventoryId,omitempty"`
	IsRelic       *bool          `json:"isRelic,omitempty"`
	League        string         `json:"league"`
	Mutated       *bool          `json:"mutated,omitempty"`
	MutatedMods   []string       `json:"mutatedMods,omitempty"`
	Name          string         `json:"name"`
	Properties    []Property     `json:"properties,omitempty"`
	Rarity        *Rarity        `json:"rarity,omitempty"`
	Requirements  []Requirement  `json:"requirements,omitempty"`
	ScourgeMods   []string       `json:"scourgeMods,omitempty"`
	Searing       *bool          `json:"searing,omitempty"`
	SecDescrText  *string        `json:"secDescrText,omitempty"`
	Shaper        *bool          `json:"shaper,omitempty"`
	SocketedItems []SocketedItem `json:"socketedItems,omitempty"`
	Sockets       []Socket       `json:"sockets,omitempty"`
	Split         *bool          `json:"split,omitempty"`
	// 仅限：宝石
	// 表示宝石是辅助技能宝石
	Support     *bool    `json:"support,omitempty"`
	Synthesised *bool    `json:"synthesised,omitempty"`
	Tangled     *bool    `json:"tangled,omitempty"`
	TypeLine    string   `json:"typeLine"`
	UtilityMods []string `json:"utilityMods,omitempty"`
	Verified    bool     `json:"verified"`
	W           int      `json:"w"`
	X           *int     `json:"x,omitempty"`
	Y           *int     `json:"y,omitempty"`
}

// Hybrid 附带技能
type Hybrid struct {
	IsVaalGem    *bool      `json:"isVaalGem,omitempty"`
	BaseTypeName string     `json:"baseTypeName"`
	Properties   []Property `json:"properties,omitempty"`
	ExplicitMods []string   `json:"explicitMods"`
	SecDescrText string     `json:"secDescrText"`
}

// SocketedItem 插槽物品
type SocketedItem struct {
	Item
	// 仅限：宝石
	Colour *SocketAttribute `json:"colour,omitempty"`
	Socket int              `json:"socket"`
}

// GetItemsResult 获取物品列表的结果
type GetItemsResult struct {
	// 使用指针以获取最佳性能
	Items     []*Item    `json:"items"`
	Character Character  `json:"character"`
	Inventory *Inventory `json:"inventory,omitempty"`
}

File: src/api/item.ts
import type { Character } from "./character.js";

export interface Inventory {
    extra_columns: number;
    /**
     * 金币
     */
    gold: number;
}

export interface Influences {
    redeemer?: boolean;
    shaper?: boolean;
    elder?: boolean;
    crusader?: boolean;
    hunter?: boolean;
    warlord?: boolean;
}

/**
 * 插槽属性，与插槽颜色一一对应。
 *
 * I: Int/智慧, D: Dex/敏捷, S: Str/力量, G: General/通用, A: Abyssal/深渊
 */
export type SocketAttribute = "I" | "D" | "S" | "G" | "A";
/**
 * 插槽颜色。
 * B: Blue/蓝色, G: Green/绿色, R: Red/红色, W: White/白色, A: Abyssal/深渊
 */
export type SocketColour = "B" | "G" | "R" | "W" | "A";

export type Rarity = "Normal" | "Magic" | "Rare" | "Unique";

export interface Socket {
    group: number;
    attr: SocketAttribute;
    sColour: SocketColour;
}

export interface Requirement {
    name: string;
    values: Array<[string, number]>;
    displayMode: number;
    type?: number;
    suffix?: string;
}

export interface Property {
    name: string;
    values: Array<[string, number]>;
    displayMode: number;
    type?: number;
    progress?: number;
}

/**
 * 物品。
 */
export interface Item {
    /**
     * 深渊珠宝
     *
     * 值为true或undefined。
     */
    abyssJewel?: boolean;
    baseType: string;
    /**
     * 已腐化
     */
    corrupted?: boolean;
    craftedMods?: string[];
    /**
     *  熔炉数据
     */
    crucible?: unknown;
    /**
     * 熔炉词缀
     */
    crucibleMods?: string[];
    /**
     * 描述文本
     */
    descrText?: string;
    /**
     * 已复制
     */
    duplicated?: boolean;
    /**
     * @deprecated 使用 influences.elder
     */
    elder?: boolean;
    /**
     * 附魔词缀
     */
    enchantMods?: string[];
    /**
     * 外延词缀
     */
    explicitMods?: string[];
    /**
     * 风味文本
     */
    flavourText?: string[];
    /**
     * 烫金传奇版本
     */
    foilVariation?: number;
    /**
     * 破溃物品
     */
    fractured?: boolean;
    /**
     * 破溃词缀
     */
    fracturedMods?: string[];
    frameType: number;
    frameTypeId: string;
    h: number;
    icon: string;
    /**
     * 嫁接物品的技能没有id
     */
    id?: string;
    /**
     * 已鉴定
     */
    identified: boolean;
    ilvl: number;
    /**
     * 孕育物品
     */
    incubatedItem?: unknown;
    /**
     * 基底词缀
     */
    implicitMods?: string[];
    /**
     * 影响
     */
    influences?: Influences;
    inventoryId?: string;
    isRelic?: boolean;
    league: string;
    /**
     * 秽生物品
     */
    mutated?: boolean;
    /**
     * 秽生词缀
     */
    mutatedMods?: string[];
    name: string;
    properties?: Property[];
    rarity?: Rarity;
    replica?: boolean;
    requirements?: Requirement[];
    /**
     * 天灾词缀
     */
    scourgeMods?: string[];
    searing?: boolean;
    /**
     * 描述文本2，仅限嫁接技能。
     */
    secDescrText?: string;
    /**
     * @deprecated 使用 influences.shaper
     */
    shaper?: boolean;
    /**
     * 插槽中的物品
     *
     * 包括：携带插槽信息的宝石、携带插槽信息的深渊珠宝、宝石（仅限嫁接物品）。
     */
    socketedItems?: ((Socketed & Gem) | (Socketed & AbyssalJewel) | Gem)[];
    sockets?: Socket[];
    /**
     * 已分裂
     */
    split?: boolean;
    synthesised?: boolean;
    tangled?: boolean;
    typeLine: string;
    /**
     * 效用词缀（比如三玉药剂）
     */
    utilityMods?: string[];
    verified: boolean;
    w: number;
    x?: number;
    y?: number;
}

/**
 * 混合技能。
 */
export interface Hybrid {
    isVaalGem?: boolean;
    baseTypeName: string;
    properties?: Property[];
    explicitMods: string[];
    secDescrText: string;
}

/**
 * 宝石/技能。
 *
 * 宝石和技能虽然概念上泾渭分明，但是在API数据中并没有严格区分。
 */
export interface Gem extends Item {
    /**
     * 是辅助？
     *
     * 只有嫁接技能(S28)没有这个字段。
     */
    support?: boolean;
    additionalProperties?: Property[];
    secDescrText: string;
    nextLevelRequirements?: Property[];
    /**
     * 部分技能会包含额外的技能。
     */
    hybrid?: Hybrid;
    /**
     * 技能可内置辅助技能。
     */
    builtInSupport?: string;
}

export interface AbyssalJewel extends Item {
    abyssJewel: boolean;
    rarity: Rarity;
}

/**
 * 插槽物品的插槽信息。
 */
export interface Socketed {
    socket: number;
    /**
     * 深渊珠宝的该属性为null
     */
    colour: SocketAttribute | null;
}

export type GetItemsResult = {
    items: Item[];
    character: Character;
    /**
     * 仅存在于获取当前Session的角色物品时
     */
    inventory?: Inventory;
};
